src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -1195,6 +1195,112 @@ void MacroAssembler::lookup_interface_method(Register recv_klass,
   }
 }
 
+// Look up the method for a megamorphic invokeinterface call in a single pass over itable:
+// - check recv_klass (actual object class) is a subtype of resolved_klass from CompiledICHolder
+// - find a holder_klass (class that implements the method) vtable offset and get the method from vtable by index
+// The target method is determined by <holder_klass, itable_index>.
+// The receiver klass is in recv_klass.
+// On success, the result will be in method_result, and execution falls through.
+// On failure, execution transfers to the given label.
+void MacroAssembler::lookup_interface_method_stub(Register recv_klass,
+                                                  Register holder_klass,
+                                                  Register resolved_klass,
+                                                  Register method_result,
+                                                  Register temp_itbl_klass,
+                                                  Register scan_temp,
+                                                  Register holder_offset,
+                                                  int itable_index,
+                                                  Label& L_no_such_interface) {
+  assert_different_registers(resolved_klass, recv_klass, holder_klass, temp_itbl_klass, scan_temp, holder_offset);
+  assert_different_registers(method_result, recv_klass, holder_klass, temp_itbl_klass, scan_temp, holder_offset);
+
+  int vtable_base = in_bytes(Klass::vtable_start_offset());
+  int itentry_off = in_bytes(itableMethodEntry::method_offset());
+  int scan_step = itableOffsetEntry::size() * wordSize;
+  int vte_size = vtableEntry::size_in_bytes();
+  int ioffset = in_bytes(itableOffsetEntry::interface_offset());
+  int ooffset = in_bytes(itableOffsetEntry::offset_offset());
+  assert(vte_size == wordSize, "else adjust times_vte_scale");
+
+  Label L_loop_scan_resolved_entry, L_resolved_found, L_holder_found;
+
+  // temp_itbl_klass = recv_klass.itable[0]
+  // scan_temp = &recv_klass.itable[0]
+  ldrw(scan_temp, Address(recv_klass, Klass::vtable_length_offset()));
+  add(recv_klass, recv_klass, vtable_base + ioffset);
+  ldr(temp_itbl_klass, Address(recv_klass, scan_temp, Address::lsl(3)));
+  mov(holder_offset, zr);
+  lea(scan_temp, Address(recv_klass, scan_temp, Address::lsl(3)));
```

Comment:
The comments to the code help little to understand the origins of the generated code.
I understand the generated code corresponds to:
```
temp_itbl_klass = (itableOffsetEntry*)(recv_klass->start_of_itable())[0].interface_klass()
```
which is
```
temp_itbl_klass = *(recv_klass + Klass::vtable_start_offset() + sizeof(intptr_t)*recv_klass->_vtable_length + itableOffsetEntry::offset_offset())
```

You store `recv_klass + Klass::vtable_start_offset() +itableOffsetEntry::offset_offset()` into `recv_klass`.
BTW, you assume `sizeof(intptr_t) == 8` to use `Address::lsl(3)`. Should this be mentioned in comments? 
You store `recv_klass + sizeof(intptr_t)*recv_klass->_vtable_length` into `scan_temp` to access `itable[i].interface_klass()`.

I think the code is worth adding more comments.

Reply:
A simple cpp code explains things better than words. Let me put it in a way that explains data access, rather than strictly following a sequence of register manipulation:
```cpp
  int vtable_start_offset = in_bytes(Klass::vtable_start_offset());
  int ioffset = in_bytes(itableOffsetEntry::interface_offset());

  ldrw(scan_temp, Address(recv_klass, Klass::vtable_length_offset()));
  add(recv_klass, recv_klass, vtable_start_offset + ioffset);
  // itableOffsetEntry[] itable = recv_klass + Klass::vtable_start_offset() + recv_klass->_vtable_len;
  // temp_itbl_klass = itable[0]._interface;
  ldr(temp_itbl_klass, Address(recv_klass, scan_temp, Address::lsl(3)));
  // scan_temp = &(itable[0]._interface)
  lea(scan_temp, Address(recv_klass, scan_temp, Address::lsl(3)));
```